File: java/GraphValidTree.java
public class Solution {
    // https://www.lintcode.com/problem/graph-valid-tree/description
    // Approach: union find to detect cycles
    private int[] unionFind;
    private int unions;
    
    private void initialize(int n) {
        unionFind = new int[n];
        for(int i = 0; i < n; ++i) {
            unionFind[i] = i;
        }
    }
    
    private int find(int node) {
        int parent;
        while(node != unionFind[node]) {
            parent = unionFind[unionFind[node]];
            unionFind[node] = parent;
            node = parent;
        }
        return node;
    }
    
    private boolean union(int left, int right) {
        int leftParent = find(left);
        int rightParent = find(right);
        
        if(leftParent == rightParent) return false;
        
        unionFind[rightParent] = leftParent;
        unions++;
        return true;
    }
    
    /**
     * @param n: An integer
     * @param edges: a list of undirected edges
     * @return: true if it's a valid tree, or false
     */
    public boolean validTree(int n, int[][] edges) {
        // write your code here
        if(edges.length == 0 && n > 1) return false;
        
        initialize(n);
        for(int[] edge : edges) {
            boolean success = union(edge[0], edge[1]);
            if(!success) return false;
        }
        
        return unions == n-1;
    }
}

File: python/graph-valid-tree.py
class Solution:
    """
    @param n: An integer
    @param edges: a list of undirected edges
    @return: true if it's a valid tree, or false
    """

    def validTree(self, n, edges):
        # This is genius act
        # imagine a tree, also this catch an issue
        # when there is two part of a graph
        if len(edges) != n - 1:
            return False

        neighbors = collections.defaultdict(list)
        for left, right in edges:
            neighbors[left].append(right)
            neighbors[right].append(left)

        # because there is no direction, so no indegree
        # visited get out of cyclone, and catch graph node not being visited
        visited = set()
        queue = collections.deque([0])

        while queue:
            node = queue.popleft()
            visited.add(node)
            for edge in neighbors[node]:
                if edge in visited:
                    continue
                visited.add(edge)
                queue.append(edge)

        return len(visited) == n

# 10
# [[0,1],[5,6],[6,7],[9,0],[3,7],[4,8],[1,8],[5,2],[5,3]]


class Solution:
    """
    @param n: An integer
    @param edges: a list of undirected edges
    @return: true if it's a valid tree, or false
    """

    def validTree(self, n, edges):
        # write your code here
        if len(edges) != n - 1:
            return False

        uf = UnionFind(n)

        for e1, e2 in edges:
            uf.merge(e1, e2)

        return uf.connected == 1


class UnionFind:
    def __init__(self, n):
        self.connected = n
        self.father = {i: i for i in range(n)}

    def find(self, n):
        path = []

        while self.father[n] != n:
            path.append(n)
            n = self.father[n]

        for item in path:
            self.father[item] = n

        return n

    def merge(self, n1, n2):
        dad1, dad2 = self.find(n1), self.find(n2)

        if dad1 != dad2:
            self.connected -= 1
            self.father[dad1] = dad2
